### ISSUE TITLE
"Exceptions"

### ISSUE THREAD

**Issue by andrew4328** [Lambda-Mountain-Compiler-Backend/lambda-mountain#994]
An "Exception" can happen in a fallible function. The result is a Monad wrapper with the exception value instead of the normal value. An infallible function must then deal with all those exception cases before it can become infallible.

```
try {
   fallible-function();
} catch MyException { x=x } => {
   handle exception
};
```

This is in-practice very different from stack-unwinding. It maybe could be implemented with stack-unwinding, but that isn't the default and isn't even encouraged.

Fallible functions do not need to be marked as fallible, this is inferred.

**Comment by andrew4328** [comment]
All Exceptions are cases of a global "AnyException" type. However, only a subset of those exceptions are expected from fallible functions.

**Comment by andrew4328** [comment]
Each Exception can have a default error message. This would permit a forced coercion to a "safe" type with a runtime error.

```
fallible-function().no-except
```

**Comment by alex-s168** [comment]
note that destructors of every value in between the failure point and the catch block have to be called!

**Comment by andrew4328** [comment]
flow typing to the rescue!

* "must be called"

**Comment by andrew4328** [comment]
There are two directions that this ticket goes too:

* softer errors with exceptions
* harder restrictions to disallow panics or "exceptional" behavior

obviously, disallowing things is easy. You just say "no".

**Comment by andrew4328** [comment]
Also there might be a misunderstanding about what an "Exception" is when comparing to a language like C++.

An exception for a function

```
let f(): Nil throws TheError = ();
```

is just syntax / type sugar for

```
let f(): Fallible<Nil,TheError> = ();
```

**Comment by alex-s168** [comment]
I just realized that we can probably attach a sticky `MightFail` tag to the `Arrow` types, to force handling of errrors

**Comment by andrew4328** [comment]
Yeah, Arrows can have associated types like `Arrow<x,y>+Fallible`

**Comment by alex-s168** [comment]
it would be cool if this worked:
```
let fail(str: String): Nil causes sticky Failable;
let getc(): I32 causes sticky Impure;
```

**Comment by alex-s168** [comment]
> Yeah, Arrows can have associated types like `Arrow<x,y>+Fallible`

acually you might want to conditionally compile code if the type of the function is **not** Failable which is impossible.

it should probably be inverted

**Comment by andrew4328** [comment]
Is "failable" already an English word? I kind of prefer it to "fallible" because it says what it means.

edit: apparently yes. This is plain English.

**Comment by andrew4328** [comment]
Syntax for annotations on functions

```
let fail(str: String): Nil causes sticky Failable;
let getc(): I32 causes sticky Impure;

is

let :Failable fail(str: String): Nil;
let :Impure getc(): I32;
```

**Comment by alex-s168** [comment]
and what about things like this:
```
let do-weird-stuff(fn: Arrow<Nil,Nil> + Not<Failable>)
```

**Comment by andrew4328** [comment]
Logical negation is not a thing in the type system.

`NotFailable` is ok and `Not<Failable>` is technically ok as long as you understand that logical connectives are ignored.

**Comment by andrew4328** [comment]
I could make negation a thing though. That example is actually pretty interesting.

**Comment by alex-s168** [comment]
additional idea:

**Comment by alex-s168** [comment]
also to consider:

when for example the fail() function is called inside an if, and it is not known at compile time if the if will succeed, it should attach information to the type that it might be Failing.

when the compiler knows that it will always call the fail() function, it should attach information to the type that is is always Failing.

if however the compiler knows that the fail() is unreachable, it should remove the Failing tag.

this should then also be extend to regular types, for example `CaseNumber<1>`
